src/systems/dynamic_detachable_joint/DynamicDetachableJoint.cc

```diff
@@ -0,0 +1,386 @@
+void DynamicDetachableJoint::PreUpdate(
+  const UpdateInfo &/*_info*/,
+  EntityComponentManager &_ecm)
+{
+  GZ_PROFILE("DynamicDetachableJoint::PreUpdate");
+  // only allow attaching if child entity is detached
+  if (this->validConfig && !this->isAttached)
+  {
+    // return if attach is not requested.
+    if (!this->attachRequested)
+    {
+      return;
+    }
+    // Look for the child model and link
+    Entity modelEntity{kNullEntity};
+
+    // if child model is the parent model
+    if ("__model__" == this->childModelName)
+    {
+      modelEntity = this->model.Entity();
+    }
+    else
+    {
+      // Querying the ECM for the child model entity
+      modelEntity = _ecm.EntityByComponents(
+          components::Model(), components::Name(this->childModelName));
+    }
+    if (kNullEntity != modelEntity)
+    {
+      this->childLinkEntity = _ecm.EntityByComponents(
+          components::Link(),
+          components::ParentEntity(modelEntity),
+          components::Name(this->childLinkName));
+
+      if (kNullEntity != this->childLinkEntity)
+      {
+        // store the child and parent link poses in the world frame
+        math::Pose3d childPose = gz::sim::worldPose(this->childLinkEntity, _ecm);
+        math::Pose3d parentPose = gz::sim::worldPose(this->parentLinkEntity, _ecm);
+
+        auto dist = childPose.Pos().Distance(parentPose.Pos());
+        gzdbg << "Centre-to-centre distance: " << dist << " m" << std::endl;
+
+        // Check if the child link is within the attach distance
+        if (dist > this->attachDistance)
+        {
+          gzwarn << "Child Link [" << this->childLinkName 
+                 << "] is too far from parent. Distance: " << dist 
+                 << "m, threshold: " << this->attachDistance << "m" << std::endl;
+          this->attachRequested = false; // reset attach request
+          return;
+        }
+        // If the child link is within the attach distance, proceed to attach
+        gzdbg << "Child Link " << this->childLinkName
+              << " is within attach distance of Parent Link. Proceeding to attach." << std::endl;
+
+        // Attach the models
+        // We do this by creating a detachable joint entity.
+        this->detachableJointEntity = _ecm.CreateEntity();
+
+        // creating the joint
+        _ecm.CreateComponent(
+            this->detachableJointEntity,
+            components::DetachableJoint({this->parentLinkEntity,
+                                         this->childLinkEntity, "fixed"}));
+        this->attachRequested = false;
+        this->isAttached = true;
+        // Keep track of the attached pair for future validation
+        this->attachedChildModelName = this->childModelName;
+        this->attachedChildLinkName  = this->childLinkName;
+        this->PublishJointState(this->isAttached);
+        gzdbg << "Attaching entity: " << this->detachableJointEntity
+              << std::endl;
+      }
+      else
+      {
```

Comment:
Knit: I recommend applying the error-first pattern (also called "happy path" programming) here. By checking for and returning early on error conditions, we can significantly reduce code nesting and enhance the overall readability.

Reference: https://preslav.me/2023/09/22/ditch-that-else/